src/compiler/emitter.ts

```diff
@@ -96,9 +96,7 @@ namespace ts {
                 comparePaths(sourceFile.fileName, ownOutputFilePath, host.getCurrentDirectory(), !host.useCaseSensitiveFileNames()) === Comparison.EqualTo;
             const jsFilePath = options.emitDeclarationOnly || isJsonEmittedToSameLocation ? undefined : ownOutputFilePath;
             const sourceMapFilePath = !jsFilePath || isJsonSourceFile(sourceFile) ? undefined : getSourceMapFilePath(jsFilePath, options);
-            // For legacy reasons (ie, we have baselines capturing the behavior), js files don't report a .d.ts output path - this would only matter if `declaration` and `allowJs` were both on, which is currently an error

```

Comment:
case when we don't write .json file because it could overwrite same json file - shoudnt write d.ts as well ?

Reply:
I don't see why `.json` and `.js` should behave any differently in that scenario.

Reply:
json files behave differently as in they do not error and don't write .json file if its going to be same path as input file.. Does that mean there should be .d.ts file generated or does it mean we need to skip the .d.ts generation as well.. In either case we need to add test to track that